crosshair/libimpl/builtinslib.py

```diff
@@ -1676,7 +1676,12 @@ def split(self, sep:Optional[str] = None, maxsplit:int = -1):
         new_maxsplit = -1 if maxsplit == -1 else maxsplit - 1
         ret.extend(self[first_occurance+1:].split(sep=sep, maxsplit=new_maxsplit))
         return ret
-
+    
+    def zfill(self, width: int):
+        size = SmtInt(z3.Length(self.var)) # Why does this need to be casted to SmtInt?
```

Comment:
Great question! Essentially:

For each Python type (say, `int`), there are two important but different other types of values that CrossHair cares about: (1) a type that emulates the built-in type (`SmtInt`) and, (2), a z3 sort (`z3.IntSort()`). The Smt* type tries very hard to be indistinguishable from the python type, while the z3 expression is just a regular z3 expression.

When crosshair analyzes a function that takes, say, an `int`, we pass in a value of type `SmtInt`. Accessing the `.var` attribute of `SmtInt` gets us to a z3 expression of sort `z3.IntSort()`. To go the other way (from z3 expression to the `SmtInt` value), we construct a new `SmtInt`, passing in a z3 expression. (it's actually a constructor, not a cast)

So, it's common to see patterns where we access `.var` on some Smt* objects, then do some work in pure z3, and then wrap that result up in some more Smt* objects.

In this case, you can fall back to some other work we've done that already interprets `len()` symbolically: this line is the same as just saying "size = len(sellf)" (and I'd probably recommend that here). For fun, look at `SmtSequence.__len__` - it is exactly your line here.